### PR TITLE
Feature/nav 174 implementation raptor xx

### PR DESCRIPTION
**Pull request by Brunner246**

### CODE REVIEW

File: Writerside/topics/raptor-xx.md [hunk 1]
```diff
@@ -1,3 +1,435 @@
-# RaptorXX
+# Simple C++ RAPTOR Implementation
 
-Implementation of the simple Raptor in C++ for performance comparison in benchmarking.
+This section discusses the design, implementation, and optimization of a public transport routing system, developed
+using modern C++20/C++23 standards. The project incorporates advanced data structures and algorithms to enable
+efficient route planning, with an emphasis on cross-platform compatibility and robust build configurations via CMake.
+
+The C++ implementation is based heavily on the [Simple Raptor](simple-raptor.md "Simple RAPTOR Implementation"), but
+extends it with optimizations and modern C++ features. In this section, we focus on the C++ implementation, detailing
```
Comment (clukas1) [inline Writerside/topics/raptor-xx.md:8]: how does it extend the Java simple raptor?

File: Writerside/topics/raptor-xx.md [hunk 2]
```diff
@@ -1,3 +1,435 @@
...
+@enduml
+```
+
+We aimed to utilize the most modern C++ features possible. Although we initially attempted to implement the project
+using the new C++ module system, limited compiler support (primarily from MSVC on Windows) posed challenges.
+
+## Challenges of Implementing in C++ Compared to Java
+
+Implementing RaptorXX in C++ presented several challenges, particularly in comparison to Java. Below, we outline key
+areas where the two languages differ, highlighting the complexities faced during C++ development.
```
Comment (clukas1) [inline Writerside/topics/raptor-xx.md:53]: General comment on this sections. I would avoid having the structure
C++: .....
Java: ....

Because most of the points you are making are: `Java (and other high level languages) natively support more utility methods, which C++ does not. To tackle this we did XY in C++.`

File: Writerside/topics/raptor-xx.md [hunk 3]
```diff
@@ -1,3 +1,435 @@
...
+
+- The CPU usage tool allows to see how much processing time the code consumes, and which parts of the code are
+  responsible for high CPU loads. This enables to focus optimization efforts on the most critical areas.
+
+- [Optimizing Code Using Profiling Tools](https://learn.microsoft.com/en-us/visualstudio/profiling/optimize-code-using-profiling-tools?view=vs-2022&source=recommendations).
+
+![cpp-diagnostic.png](cpp-diagnostic.png)(compiler_support.png){width="850"}
+![cpp-diagnostic-scanRoute.png](cpp-diagnostic-scanRoute.png){width="850"}
+
+## Foreign Function and Memory (FFM) API
```
Comment (clukas1) [inline Writerside/topics/raptor-xx.md:360]: I think this section does not really explain why we didn't implement the FFM API. The reason was that sending routing requests from JAVA to C++, relied on C++ being able to operate on the schedule memory layout built in JAVA. Since JAVA would be the only active application, alternatively building the schedule from files in C++ from files for each request (not performant at all). The better solution would have been having a C++ service, which Java could address when needed for routing.

Summary, FFM API is not suited for the application we were developing. It's better to replace complex functions instead of whole packages.

**Review by clukas1 — CHANGES_REQUESTED** [review]

File: Writerside/topics/raptor-xx.md (hunk 1 above)
Comment (Brunner246) [reply Writerside/topics/raptor-xx.md:8]: That was meant more in the sense that it adopts the same logic.

File: Writerside/topics/raptor-xx.md (hunk 2 above)
Comment (Brunner246) [reply Writerside/topics/raptor-xx.md:53]: Would you like to have the section removed completely?

**Comment by Brunner246** [comment]
info, i will add next to the time measure table the system info on what ma machine i ran the tests.
@Brunner246

File: Writerside/topics/raptor-xx.md [hunk 4]
```diff
@@ -47,81 +47,76 @@ package "src" {
...
-  offer competitive performance, though not always at the same level of fine-tuned C++ code.
-
-### Conclusion
-
-Developing a public transit routing system like RaptorXX in C++ presents unique challenges compared to Java,
-especially concerning cross-platform compatibility, library support, memory management, and build complexity. While C++
-offers more granular control and performance potential, it requires greater effort in managing these aspects. In
-contrast, Java's comprehensive standard library, automatic memory management, and mature ecosystem can significantly
-simplify development.
+## Challenges of Implementing in C++
```
Comment (clukas1) [inline Writerside/topics/raptor-xx.md:50]: No, I didn't actually mean to remove the entire Java comparison. My idea was to make the text more natural with less bullet point.

Here a (semi-me/semi chatgpt) reworked version I would suggest:

## Challenges of Implementing in C++

Implementing RaptorXX in C++ aimed to improve performance over Java, but this proved to be a challenging task. Java's JVM is highly optimized, often providing excellent performance through Just-In-Time (JIT) compilation and various built-in optimization strategies. In contrast, C++ demands careful attention to several complex areas, which are discussed below.

### Cross-Platform Compatibility
Achieving cross-platform compatibility in C++ is much more intricate than in Java. Java’s “write once, run anywhere” approach, thanks to the JVM, abstracts away many platform differences. In C++, developers must contend with variations in compilers, libraries, and system calls across different operating systems like Windows, Linux, and macOS. This often requires using conditional compilation and platform-specific code. 

While the C++ Standard Library (STL) provides essential functionality, it lacks built-in support for common tasks like logging or CSV parsing. By comparison, Java has robust libraries like `java.util.logging` and `java.nio.file` readily available. In C++, you often need to integrate third-party libraries such as `spdlog` for logging or `csv-parser` for CSV handling, which adds complexity to the project. Furthermore, not all compilers support the full range of C++20 features, requiring careful attention to feature compatibility for cross-platform consistency.

### Memory Management
While Java handles memory management through automatic garbage collection, C++ requires manual memory management, which can be error-prone. Memory leaks, buffer overflows, and dangling pointers are common pitfalls that developers must avoid. 

Modern C++ does offer tools to mitigate these risks, such as smart pointers (`std::unique_ptr`, `std::shared_ptr`), but their proper use requires careful planning. Java developers rarely face these issues, as memory management is largely abstracted away by the JVM, allowing them to focus on higher-level design concerns. In C++, developers need a deep understanding of memory allocation to ensure both safety and performance.

### Error Handling
Error handling in C++ is typically done through exceptions, similar to Java, but ensuring exception safety and proper resource cleanup can be more difficult in C++. 

Java’s `try-catch` blocks and garbage collection simplify this process, whereas in C++, developers must manually ensure that all resources are released, especially when exceptions are thrown. This is where tools like RAII (Resource Acquisition Is Initialization) come into play, but it requires careful design to implement correctly.

### Dependency and Build Management
Managing dependencies and builds in C++ is another area where complexity arises compared to Java. Java’s ecosystem offers Maven and Gradle, which are tightly integrated with the development environment, making dependency management straightforward. 

In C++, tools like CMake are available to help with build management, but they often require extensive configuration, especially for larger projects. We relied on `vcpkg` from Microsoft for handling dependencies, but tools like `vcpkg` and Conan, while useful, are not as seamless as Java’s ecosystem. 

Additionally, setting up development tools such as debuggers and profilers is generally more labor-intensive in C++, requiring greater familiarity with the toolchain. In our case, we primarily used CLion from JetBrains but occasionally had to switch to Visual Studio for certain tasks, adding further complexity to the workflow.

### Performance Optimization
One of C++'s primary strengths is the fine-grained control it offers over system performance, but taking full advantage of this requires deep knowledge of both the language and hardware architecture. In contrast, Java abstracts away most of these details, allowing developers to focus more on application logic while JVM optimizations handle much of the performance tuning behind the scenes. 

While C++ can theoretically outperform Java due to its lower-level access, writing optimized C++ code that is both maintainable and readable requires a delicate balance. This trade-off often makes development slower and more challenging in C++ than in Java.

---

In summary, while C++ offers greater control and potential for higher performance, the trade-offs in complexity, especially in areas like cross-platform support, memory management, and build tools, make it a significantly more challenging language to work with compared to Java.

File: Writerside/topics/raptor-xx.md [hunk 5]
```diff
@@ -47,81 +47,76 @@ package "src" {
...
+but are not as integrated or straightforward as Java’s ecosystem. We relied on `vcpkg` from Microsoft. The C++ ecosystem
+offers a wide range of tools, though their quality and ease of use can vary. Setting up debugging, profiling, and other
+development tools often requires more effort. We mainly worked with CLion from JetBrains but had to switch to Visual
+Studio for certain tasks.
+
+While C++ offers fine-grained control over system performance, leveraging these optimizations requires deep knowledge of
+both the language and hardware architecture. Developers must often balance performance with code maintainability and
+readability.
+
+### Conclusion - JAVA vs C++ Performance
```
Comment (clukas1) [inline Writerside/topics/raptor-xx.md:80]: would raise the level of this header to two #

File: Writerside/topics/raptor-xx.md [hunk 6]
```diff
@@ -47,81 +47,76 @@ package "src" {
...
+|----------------------------------------|-------------------------------------|------------|------------------------|-----------------------|-----------------|----------------|
+| 8589640 (St. Gallen, Vonwil)           | 8579885 (Mels, Bahnhof)             | 100        | 13                     | 19                    | -6              | -31.6%         |
+| 8574563 (Maienfeld, Bahnhof)           | 8587276 (Biel/Bienne, Taubenloch)   | 100        | 55                     | 38                    | 17              | 44.7%          |
+| 8588524 (Sion, Hôpital Sud)            | 8508896 (Stans, Bahnhof)            | 100        | 45                     | 20                    | 25              | 125.0%         |
+| 8510709 (Lugano, Via Domenico Fontana) | 8579255 (Lausanne, Pont-de-Chailly) | 100        | 46                     | 29                    | 17              | 58.6%          |
+| 8574848 (Davos Dorf, Bahnhof)          | 8576079 (Rapperswil SG, Sonnenhof)  | 100        | 12                     | 15                    | -3              | -20.0%         |
```
Comment (clukas1) [inline Writerside/topics/raptor-xx.md:106]: maybe add average difference?

File: Writerside/topics/raptor-xx.md [hunk 7]
```diff
@@ -360,48 +347,16 @@ helping you identify and resolve performance bottlenecks and resource management
...
-
-Although the FFM API abstracts some platform-specific details, developers must still account for differences in calling
-conventions and memory layouts between platforms (e.g., Windows vs. Linux). These differences can introduce subtle bugs
-or performance issues when deploying the same Java code across different operating systems. Cross-platform compatibility
-must be thoroughly tested to avoid such pitfalls.
+Memory (FFM) API. However, we faced significant challenges due to the complexity of mapping intricate data structures
+between C++ and Java. Ultimately, we decided against this approach.
 
-### Conclusion (FFM)
+The FFM API was not well-suited for our application. Sending routing requests from Java to C++ required C++ to operate
```
Comment (clukas1) [inline Writerside/topics/raptor-xx.md:353]: Suggestion:

The FFM (Foreign Function & Memory) API wasn't well-suited for our application. Sending routing requests from Java to C++ required C++ to interact with a memory structure built in Java, which would have been complex to implement. Since Java was the primary application, the alternative was rebuilding the schedule from files in C++ for each request, which would have been extremely inefficient.

A more effective approach would have been to implement a dedicated C++ service that Java could call for routing tasks. The FFM API, while useful for certain types of function replacements, wasn't designed to manage the kind of larger, integrated workflow we needed, making it unsuitable for our use case.

**Review by clukas1 — CHANGES_REQUESTED** [review]
Nice work, I like how the benchmarking is coming in. Some minor suggestions from my side.

File: Writerside/topics/raptor-xx.md (hunk 4 above)
Comment (Brunner246) [reply Writerside/topics/raptor-xx.md:50]: Thanks for the adjustments. I have now used a mixture of your text and mine again.

**Review by clukas1 — CHANGES_REQUESTED** [review]
Looks good, only change I would do is move the benchmarking results to the results/discussion part and reference them in the implementation.

**Review by clukas1 — APPROVED** [review]
